contributing/guide-observability.md

```diff
@@ -120,6 +120,63 @@ logs don't show up in production normally.
+- `cannot resolve package dependencies in crossplane.yaml: invalid Semantic Version "main" on package "provider-kubernetes"`
+- `failed to load OCI image provider-gcp/v0.42.0: failed to unpack layer: cannot parse file: failed to parse YAML of file foo.yaml: unexpected indention in line 4`.
+
+General rules:
+1. follow the pattern `generic error cause: more details: even more details`.
+2. put values into quotes (`%q` in Golang), with the exception of
+   - resource names with namespace or resource type (e.g. `default/foo` or `providers/foo`), resource names alone are quoted
+   - kinds (e.g. `cannot find provider "provider-kubernetes"`; here, the kind provider is without quotes, the provider name is with)
+   - filenames with extensions and/or path (e.g. `cannot parse foo.yaml`).
+3. add a context of e.g. filenames, object names, or values which are wrong (and the values are known to be insensitive).
+
+   Which context is added by which functions depends on the level of abstraction of the code at hand: the package loader will know the OCI image name, the YAML parser might only see a byte slice. The former adds the image name as context, some layer in-between adds the filename and the YAML parser adds the line and column in the byte stream.
+
+   Filenames should be relative to the logical working directory of the context at hand (e.g. a package root, current working directory, workspace root). Ensure that the context is clear in the error message (e.g. `failed to load package provider-kubernetes: failed to parse apis/xrd.yaml`).
+
+4. add a context just once.
+
+   Rule of thumb: if a function gets a parameter, it will also include it in the error messages, so the caller does not have to.
+   
+   E.g. a function getting a filename as input will include the filename in errors it returns.
+5. don't return multi-line errors.
+6. aggregate errors if there is clear use for the user, but not in general, i.e. fail fast is the default.
+
+   Aggregate through `kerrors.NewAggregate`. When the number of aggregated errors is unbounded,
+   return a subset, e.g. the first 3.
+
+   E.g.: `cannot parse workspace: failed to parse foo.yaml, bar.yaml, abc.yaml, and 17 more files`
+7. don't include sensitive information like tokens into error messages, not even in debug mode.
+
+   These are usually considered insensitive: field names, file paths, resources, kinds, object names, namespaces, labels, enum values, numbers, booleans.
+8. use error message constants, e.g. `errSomethingWentWrong` for messages without interpolation, and `errFmtSomethingWentWrong` for those with interpolation.
+9. use [wrapping errors] and github.com/crossplane/crossplane-runtime/pkg/errors in general.
+10. error messages must be deterministic. Use sorting in loops over maps or sort returned errors to ensure determinism.
+
+    Background: one never knows how an error is displayed. Not being deterministic can lead to hotloops
+    with constant API updates, e.g. of a condition.
```

Comment:
👍 FWIW there are concrete examples of this having been an issue for us. For example any AWS errors have unique IDs that change every time the same error occurs, causing constant condition updates.